File: plugin/hooks/get-read-pretool-output.py
#!/usr/bin/env python3
"""
get-read-pretool-output.py - Unified PreToolUse hook for Read/Glob/Grep

TRIGGER: PreToolUse (matcher: Read|Glob|Grep)

Consolidates read operation validation hooks into a single Python dispatcher.

Handlers can:
- Warn about patterns (return {"warning": "..."})
- Block operations (return {"block": True, "message": "..."})
- Allow silently (return None)

Benefits:
1. Single Python startup vs N shell startups
2. Native JSON parsing (no jq subprocesses)
3. Shared state management
4. Better error handling
"""

import json
import sys
from pathlib import Path

# Add handlers directory to path
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from read_pretool_handlers import get_all_handlers


def read_hook_input() -> dict:
    """Read and parse JSON input from stdin."""
    if sys.stdin.isatty():
        return {}

    try:
        raw = sys.stdin.read()
        return json.loads(raw) if raw.strip() else {}
    except json.JSONDecodeError:
        return {}


def main():
    # Read hook input
    hook_data = read_hook_input()

    # Extract tool info
    tool_name = hook_data.get("tool_name", "")
    if tool_name not in ("Read", "Glob", "Grep"):
        print("{}")
        return

    tool_input = hook_data.get("tool_input", {})

    # Build context for handlers
    context = {
        "tool_input": tool_input,
        "session_id": hook_data.get("session_id", ""),
        "hook_data": hook_data,
    }

    # Run all handlers
    warnings = []
    for handler in get_all_handlers():
        try:
            result = handler.check(tool_name, context)
            if result:
                if result.get("block"):
                    # Block the operation
                    print(json.dumps({
                        "decision": "block",
                        "reason": result.get("message", "Blocked by handler")
                    }))
                    return
                if result.get("warning"):
                    warnings.append(result["warning"])
        except Exception as e:
            sys.stderr.write(f"get-read-pretool-output: handler error: {e}\n")

    # Output warnings if any
    if warnings:
        for warning in warnings:
            sys.stderr.write(warning + "\n")

    # Allow the operation
    print("{}")


if __name__ == "__main__":
    main()

File: plugin/hooks/src/io/github/cowwoc/cat/hooks/GetReadPretoolOutput.java
package io.github.cowwoc.cat.hooks;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import tools.jackson.databind.JsonNode;

/**
 * get-read-pretool-output - Unified PreToolUse hook for Read/Glob/Grep
 *
 * TRIGGER: PreToolUse (matcher: Read|Glob|Grep)
 *
 * Consolidates read operation validation hooks into a single Java dispatcher.
 *
 * Handlers can:
 * - Warn about patterns (return warning)
 * - Block operations (return block=true with message)
 * - Allow silently (return null)
 */
public final class GetReadPretoolOutput
{
  private static final Set<String> SUPPORTED_TOOLS = Set.of("Read", "Glob", "Grep");

  /**
   * Entry point for the Read pretool output hook.
   *
   * @param args command line arguments (unused)
   */
  public static void main(String[] args)
  {
    HookInput input = HookInput.readFromStdin();

    String toolName = input.getToolName();
    if (!SUPPORTED_TOOLS.contains(toolName))
    {
      HookOutput.empty();
      return;
    }

    JsonNode toolInput = input.getToolInput();
    String sessionId = input.getSessionId();
    List<String> warnings = new ArrayList<>();

    // Run all read pretool handlers
    for (ReadHandler handler : HandlerRegistry.getReadPretoolHandlers())
    {
      try
      {
        ReadHandler.Result result = handler.check(toolName, toolInput, null, sessionId);
        if (result.blocked())
        {
          // Handler blocked the operation
          HookOutput.block(result.reason(), result.additionalContext());
          return;
        }
        if (result.reason() != null && !result.reason().isEmpty())
        {
          warnings.add(result.reason());
        }
      }
      catch (Exception e)
      {
        System.err.println("get-read-pretool-output: handler error: " + e.getMessage());
      }
    }

    // Output warnings if any
    for (String warning : warnings)
    {
      System.err.println(warning);
    }

    // Allow the operation
    HookOutput.empty();
  }
}
